src/Core/Datatypes/DyadicTensor.h

```diff
@@ -0,0 +1,322 @@
+   and/or sell copies of the Software, and to permit persons to whom the
+   Software is furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included
+   in all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef CORE_DATATYPES_DYADIC_TENSOR_H
+#define CORE_DATATYPES_DYADIC_TENSOR_H
+
+#include <Core/Datatypes/TensorBase.h>
+#include <Core/Datatypes/TensorFwd.h>
+#include <Core/Utils/Exception.h>
+#include <Eigen/Dense>
+#include <algorithm>
+#include <initializer_list>
+#include <limits>
+#include <Core/Datatypes/share.h>
+
+namespace SCIRun {
+namespace Core {
+  namespace Datatypes {
+    // Dyadic tensors are also known as second-order tensors
+    template <typename Number, size_t Dim>
+    class DyadicTensorGeneric : public TensorBaseGeneric<Number, Eigen::Sizes<Dim, Dim>>
+    {
+     public:
+      using parent = TensorBaseGeneric<Number, Eigen::Sizes<Dim, Dim>>;
+      using VectorType = Eigen::Matrix<Number, Dim, 1>;
+      using MatrixType = Eigen::Matrix<Number, Dim, Dim>;
+
+      DyadicTensorGeneric() : parent() { parent::setZero(); }
+
+      DyadicTensorGeneric(Number val) : parent()
+      {
+        for (size_t i = 0; i < Dim; ++i)
+          for (size_t j = 0; j < Dim; ++j)
+            (*this)(i, j) = (i == j) ? val : 0;
+      }
+
+      DyadicTensorGeneric(const std::vector<VectorType>& eigvecs) : parent()
+      {
+        if (eigvecs.size() != Dim)
+          THROW_INVALID_ARGUMENT("The number of input vectors must be " + Dim);
+        setEigenVectors(eigvecs);
+      }
+
+      DyadicTensorGeneric(const std::initializer_list<VectorType>& eigvecs) : parent()
+      {
+        if (eigvecs.size() != Dim)
+          THROW_INVALID_ARGUMENT("The number of input vectors must be " + Dim);
+        setEigenVectors(eigvecs);
+      }
+
+      DyadicTensorGeneric(const DyadicTensorGeneric<Number, Dim>& other) : parent()
+      {
+        for (size_t i = 0; i < Dim; ++i)
+          for (size_t j = 0; j < Dim; ++j)
+            (*this)(index(i), index(j)) = other(index(i), index(j));
+        eigvecs_ = other.getEigenvectors();
+        eigvals_ = other.getEigenvalues();
+        haveEigens_ = true;
+      }
+
+      DyadicTensorGeneric(DyadicTensorGeneric<Number, Dim>&& other) : parent()
+      {
+        for (size_t i = 0; i < Dim; ++i)
+          for (size_t j = 0; j < Dim; ++j)
+            (*this)(index(i), index(j)) = std::move(other(index(i), index(j)));
```

Comment:
This will just move individual doubles, or floats--not really worth it. Can you do a larger bulk move of the whole internals? Does the Eigen base class have a faster move constructor?